processor/src/main/java/org/mapstruct/ap/model/PropertyMapping.java

```diff
@@ -130,23 +138,33 @@ public PropertyMappingBuilder existingVariableNames(Collection<String> existingV
         public PropertyMapping build() {
 
             // handle target
-            TargetAccessorType targetAccessorType = getTargetAcccessorType();
+            TargetWriteAccessorType targetAccessorType = getTargetAcccessorType();
             Type targetType = getTargetType( targetAccessorType );
 
             // handle source
             String sourceElement = getSourceElement();
             Type sourceType = getSourceType();
             String sourceRefStr;
-            if ( targetAccessorType == TargetAccessorType.ADDER && sourceType.isCollectionType() ) {
+            if ( targetAccessorType == TargetWriteAccessorType.ADDER && sourceType.isCollectionType() ) {
                 // handle adder, if source is collection then use iterator element type as source type.
                 // sourceRef becomes a local variable in the itereation.
                 sourceType = sourceType.getTypeParameters().get( 0 );
-                sourceRefStr = Executables.getElementNameForAdder( targetAccessor );
+                sourceRefStr = Executables.getElementNameForAdder( targetWriteAccessor );
             }
             else {
                 sourceRefStr = getSourceRef();
             }
 
+            // all the tricky cases will be excluded for the time being.
```

Comment:
Could you open a follow-up issue describing those? It's not that we need to implement all of them necessarily, but good to keep a record of what you consider to be missing.